README.md
# mev-relay

This repository contains a public relay for accepting transactions from searchers. It also contains
an example reverse proxy for miners to run in front of their mev-geth nodes. This relay is meant
only to protect participating miners from abuse via DoS attacks, but does otherwise no bundle
filtering or censoring.

## Public Relay

[This is the relay entrypoint](server/main.js). The public flashbots relay is available at
https://relay.flashbots.net. See https://github.com/flashbots/ethers-provider-flashbots-bundle for a
library to help you call this.

The relay provides new JSON-RPC methods for interfacing with Flashbots. They are documented below:

### eth_sendBundle

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "eth_sendBundle",
  "params": [{ txs, blockNumber, minTimestamp, maxTimestamp, revertingTxHashes }]
}
```

- **txs**: Array[String], A list of signed transactions to execute in an atomic bundle
- **blockNumber**: String, a hex encoded block number for which this bundle is valid on
- **minTimestamp(Optional)**: Number, the minimum timestamp for which this bundle is valid, in
  seconds since the unix epoch
- **maxTimestamp(Optional)**: Number, the minimum timestamp for which this bundle is valid, in
  seconds since the unix epoch
- **revertingTxHashes(Optional)**: Array[String], list of tx hashes within the bundle that are
  allowed to revert

Example:

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "eth_sendBundle",
  "params": [
    {
      "txs": ["0x123abc...", "0x456def..."],
      "blockNumber": "0xb63dcd",
      "minTimestamp": 0,
      "maxTimestamp": 1615920932
    }
  ]
}
```

### eth_callBundle

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "eth_callBundle",
  "params": [{ txs, blockNumber, stateBlockNumber, timestamp }]
}
```

- **txs**: Array[String], A list of signed transactions to execute in an atomic bundle
- **blockNumber**: String, a hex encoded block number for which this bundle is valid on
- **stateBlockNumber**: String, either a hex encoded number or a block tag for which state to base
  this simulation on. Can use "latest"
- **timestamp(Optional)**: Number, the timestamp to use for this bundle simulation, in seconds since
  the unix epoch

Example:

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "eth_callBundle",
  "params": [
    {
      "txs": ["0x123abc...", "0x456def..."],
      "blockNumber": "0xb63dcd",
      "stateBlockNumber": "latest",
      "timestamp": 1615920932
    }
  ]
}
```

### flashbots_getUserStats

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "flashbots_getUserStats",
  "params": [blockNumber]
}
```

- **blockNumber**: String, a hex encoded recent block number, in order to prevent replay attacks.
  Must be within 20 blocks of the current chain tip.

Returns a quick summary of how this searcher is performing in the relay. Currently it is updated
once every hour.

Example response:

```json
{
  "is_high_priority": true,
  "all_time_miner_payments": "1280749594841588639",
  "all_time_gas_simulated": "30049470846",
  "last_7d_miner_payments": "1280749594841588639",
  "last_7d_gas_simulated": "30049470846",
  "last_1d_miner_payments": "142305510537954293",
  "last_1d_gas_simulated": "2731770076"
}
```

- **is_high_priority**: boolean representing if this searcher has a high enough reputation to be in
  the high priority queue
- **all_time_miner_payments**: The total amount paid to miners over all time
- **all_time_gas_simulated**: The total amount of gas simulated across all bundles submitted to the
  relay. This is the actual gas used in simulations, not gas limit

## Authentication

This relay requires that all payloads are signed with an ethereum wallet.

The signature is calculated by taking the EIP-191 hash of the json body encoded as UTF-8 bytes.
Here's an example using ethers.js:

```js
body = '{"id": 1234, "method", "eth_sendBundle", "params": [["0x123..."], "0xB84969"]}';
wallet = ethers.Wallet.createRandom();
wallet.signMessage(ethers.utils.id(body));
```

or in web3py:

```py
from web3 import Web3
from eth_account import Account, messages

body = '{"id": 1234, "method", "eth_sendBundle", "params": [["0x123..."], "0xB84969"]}'
message = messages.encode_defunct(text=Web3.keccak(text=body).hex())
signed_message = Account.sign_message(message, private_key=private_key_hex)
```

or in go:

```go
hashedBody := crypto.Keccak256Hash([]byte(body)).Hex()
sig, err := crypto.Sign(crypto.Keccak256([]byte("\x19Ethereum Signed Message:\n"+strconv.Itoa(len(hashedBody))+hashedBody)), pk)
signature := addr.Hex() + ":" + hexutil.Encode(sig)
```

Take this signature and append it to the ethereum address of the signer, separated by a colon, `:`.
Then send it in the `X-Flashbots-Signature` HTTP header like so:

```
X-Flashbots-Signature: 0x95c622A2c597a8bdC26D371Dd3D57dA9D26052DF:0xc73d4790fed41954869625c159a4617e3374019839a8ad72de15e41371719d6873c780e00293fcdc100aa505f33dd8480e7b07551483c8c438fe8236972d26ca1c
```

This signer does not have to be related to the signer of your actual transactions. It is just used
for authentication/rate limiting purposes, and is how `flashbots_getUserStats` determines the user.

## Miners

See [https://github.com/flashbots/mev-proxy](https://github.com/flashbots/mev-proxy) for an example
reverse proxy that this relay can connect to. Also, take a look at
[https://github.com/flashbots/mev-geth](https://github.com/flashbots/mev-geth)


CHANGELOG.md
# Changelog

# v3.0.1 (2021-07-28)

- Nodejs packaging improvements

## v3.0.0 (2021-05-06)

- Update for v0.2 core

## v2.1.0 (2021-03-25)

- Bundles are now required to send at least 1 wei to coinbase

## v2.0.0 (2021-03-03)

- Add X-Flashbots-Signature as an alternative to api-key Authorization header
- Add SQS queuing instead of directly calling miners
- Remove global rate limit

## v1.0.0 (2021-01-02)

- Initial release


miner/proxy.js
// A simple server that proxies only specific methods to an Ethereum JSON-RPC
const express = require('express');
const bodyParser = require('body-parser');
const request = require('request');
const _ = require('lodash');

const ALLOWED_METHODS = ['eth_sendBundle'];

function help() {
  console.log('node ./miner/proxy.js [PUBLIC_PORT] [GETH_PORT] [GETH_URL]');
}

function validPort(port) {
  if (isNaN(port) || port < 0 || port > 65535) {
    return false;
  }

  return true;
}

if (_.includes(process.argv, '-h') || _.includes(process.argv, '--help')) {
  help();
  process.exit(0);
}

const PUBLIC_PORT = parseInt(_.get(process.argv, '[2]', '18545'));
const GETH_PORT = parseInt(_.get(process.argv, '[3]', '8545'));
const GETH_URL = _.get(process.argv, '[4]', 'http://localhost');

if (!validPort(PUBLIC_PORT)) {
  console.error(`invalid port specified for PUBLIC_PORT: ${PUBLIC_PORT}`);
  process.exit(1);
}
if (!validPort(GETH_PORT)) {
  console.error(`invalid port specified for GETH_PORT: ${GETH_PORT}`);
  process.exit(1);
}

const app = express();
app.use(bodyParser.json());

app.use(function (req, res) {
  if (!req.body) {
    res.writeHead(400);
    res.end('invalid json body');
    return;
  }
  if (!req.body.method) {
    res.writeHead(400);
    res.end('missing method');
    return;
  }
  if (!_.includes(ALLOWED_METHODS, req.body.method)) {
    res.writeHead(400);
    res.end(`invalid method, only ${ALLOWED_METHODS} supported, you provided: ${req.body.method}`);
    return;
  }

  request
    .post({
      url: `${GETH_URL}:${GETH_PORT}`,
      body: JSON.stringify(req.body),
      headers: { 'Content-Type': 'application/json' },
    })
    .on('error', function (e) {
      res.writeHead(500);
      res.end(`error in proxy: ${e}`);
    })
    .pipe(res);
});

app.listen(PUBLIC_PORT, () => {
  console.log(`proxy listening at ${PUBLIC_PORT} and forwarding to ${GETH_URL}:${GETH_PORT}`);
});


server/bundle.js
const { ethers } = require('ethers');
const _ = require('lodash');
const { arrayify, keccak256 } = require('ethers/lib/utils');
const util = require('util');

const BLACKLIST = [
  // OFAC banned addresses
  '0x8576acc5c05d6ce88f4e49bf65bdf0c62f91353c',
  '0xd882cfc20f52f2599d84b8e8d58c7fb62cfe344b',
  '0x901bb9583b24d97e995513c6778dc6888ab6870e',
  '0xa7e5d5a720f06526557c513402f2e6b5fa20b00', // this is an invalid address, but is what's listed in the OFAC ban list
  '0xa7e5d5a720f06526557c513402f2e6b5fa20b008', // the actual valid address
  '0x7f367cc41522ce07553e823bf3be79a889debe1b',
];

function checkBlacklistTx(tx) {
  return (
    (tx.to && _.includes(BLACKLIST, tx.to.toString())) ||
    (tx.from && _.includes(BLACKLIST, tx.from.toString()))
  );
}

function checkBlacklist(txs) {
  for (let i = 0; i < txs.length; i++) {
    const tx = txs[i];
    if (checkBlacklistTx(tx)) {
      return true;
    }
  }

  return false;
}

function getParsedTransactions(rawTxs) {
  return rawTxs.map((rawTx) => {
    return ethers.utils.parseTransaction(rawTx);
  });
}

function generateBundleHash(txs) {
  let hashes = '0x';
  for (let i = 0; i < txs.length; i++) {
    const tx = txs[i];
    hashes += tx.hash().toString('hex');
  }

  return keccak256(hashes);
}
module.exports = { checkBlacklist, getParsedTransactions, generateBundleHash };


server/handlers.js
const fetch = require('node-fetch');
const Sentry = require('@sentry/node');
const AWS = require('aws-sdk');
const postgres = require('postgres');
const util = require('util');

const { writeError } = require('./utils');
const { checkBlacklist, getParsedTransactions, generateBundleHash } = require('./bundle');

const MIN_GAS_FLOOR = 42000;
function convertBundleFormat(bundle) {
  if (!Array.isArray(bundle[0])) {
    // is already v2 bundle, just return
    bundle[0].version = 2;
    return bundle[0];
  }

  const newBundle = {
    txs: bundle[0],
    blockNumber: bundle[1],
  };

  if (bundle[2]) {
    newBundle.minTimestamp = bundle[2];
  }

  if (bundle[3]) {
    newBundle.maxTimestamp = bundle[3];
  }

  return newBundle;
}
function convertSimBundleFormat(bundle) {
  if (!Array.isArray(bundle[0])) {
    return bundle[0];
  }

  const newBundle = {
    txs: bundle[0],
    blockNumber: bundle[1],
    stateBlockNumber: bundle[2],
  };

  if (bundle[3]) {
    newBundle.timestamp = bundle[3];
  }

  return newBundle;
}

class Handler {
  constructor(MINERS, promClient) {
    this.MINERS = MINERS;
    this.bundleCounter = new promClient.Counter({
      name: 'bundles',
      help: '# of bundles received',
    });
  }

  async handleSendBundle(req, res) {
    if (!req.body.params || !req.body.params[0]) {
      writeError(res, 400, 'missing params');
      return;
    }
    this.bundleCounter.inc();
    const bundle = convertBundleFormat(req.body.params);
    req.body.params = [bundle];

    const txs = bundle.txs;
    let bundleHash;

    try {
      const parsedTransactions = getParsedTransactions(txs);

      if (checkBlacklist(parsedTransactions)) {
        console.error(`txs was interacting with blacklisted address: ${txs}`);
        writeError(res, 400, 'blacklisted tx');
        return;
      }
      bundleHash = generateBundleHash(parsedTransactions);
    } catch (error) {
      console.error(`error decoding bundle: ${error}`);
      writeError(res, 400, 'unable to decode txs');
      return;
    }
    const blockParam = bundle.blockNumber;
    if (!blockParam) {
      writeError(res, 400, 'missing block param');
      return;
    }
    if (blockParam.slice(0, 2) !== '0x' || !(parseInt(blockParam, 16) > 0)) {
      writeError(res, 400, 'block param must be a hex int');
      return;
    }
    const minTimestamp = bundle.minTimestamp;
    if (minTimestamp && !(minTimestamp > 0)) {
      writeError(res, 400, 'minTimestamp must be an int');
      return;
    }
    const maxTimestamp = bundle.maxTimestamp;
    if (maxTimestamp && !(maxTimestamp > 0)) {
      writeError(res, 400, 'maxTimestamp must be an int');
      return;
    }

    const requests = [];

    console.log('req.body', req.body);
    this.MINERS.forEach((minerUrl) => {
      try {
        requests.push(
          fetch(`${minerUrl}`, {
            method: 'post',
            body: JSON.stringify(req.body),
            headers: { 'Content-Type': 'application/json' },
          }),
        );
      } catch (error) {
        Sentry.captureException(error);
        console.error('Error calling miner', minerUrl, error);
      }
    });

    if (this.SQS_URL) {
      const params = {
        DelaySeconds: 0,
        MessageAttributes: {},
        MessageBody: JSON.stringify(req.body),
        QueueUrl: this.SQS_URL,
      };
      if (req.user.keyID) {
        params.MessageAttributes.KeyID = {
          DataType: 'String',
          StringValue: req.user.keyID,
        };
      }
      if (req.user.address) {
        params.MessageAttributes.SignerAddress = {
          DataType: 'String',
          StringValue: req.user.address,
        };
      }

      await this.sqs.sendMessage(params).promise();
    }
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify({ jsonrpc: '2.0', id: req.body.id, result: { bundleHash } }));
  }

  async handleCallBundle(req, res) {
    const bundle = convertSimBundleFormat(req.body.params);
    bundle.coinbase = process.env.COINBASE_ADDRESS;
    req.body.params = [bundle];

    const txs = bundle.txs;
    const parsedTransactions = getParsedTransactions(txs);
    try {
      if (checkBlacklist(parsedTransactions)) {
        console.error(`bundle was interacting with blacklisted address: ${parsedTransactions}`);
        writeError(res, 400, 'blacklisted tx');
        return;
      }
    } catch (error) {
      console.error(`error decoding bundle: ${error}`);
      writeError(res, 400, 'unable to decode txs');
      return;
    }

    try {
      const resp = await fetch(this.SIMULATION_RPC, {
        method: 'POST',
        body: JSON.stringify(req.body),
        headers: { 'Content-Type': 'application/json' },
      });

      const result = await resp.json();
      if (result.result) {
        if (result.result.totalGasUsed < MIN_GAS_FLOOR) {
          writeError(res, 400, `bundle used too little gas, must use at least ${MIN_GAS_FLOOR}`);
        }
      }

      res.json(result);
    } catch (error) {
      console.error(`error simulating bundle: ${error}`);
      writeError(res, 400, 'failed to simulate');
    }
  }

  async handleUserStats(req, res) {
    if (req.user.keyID) {
      const stats = await this.sql`
      select
          *
      from
        stats_by_user_key_id
      where
          ${req.user.keyID} = user_key_id`;

      if (stats.length === 0) {
        return res.json({ error: { message: "stats don't exist for this user", code: -32602 } });
      }
      res.json({ result: stats[0] });
    } else {
      const stats = await this.sql`
      select
          *
      from
        stats_by_signing_address
      where
          ${req.user.address} = signing_address`;

      if (stats.length === 0) {
        return res.json({ error: { message: "stats don't exist for this user", code: -32602 } });
      }
      res.json({ result: stats[0] });
    }
  }
}

module.exports.Handler = Handler;


server/main.js
// A simple server that proxies only specific methods to an Ethereum JSON-RPC
const express = require('express');
const bodyParser = require('body-parser');
const morgan = require('morgan');
const rateLimit = require('express-rate-limit');
const _ = require('lodash');
const Sentry = require('@sentry/node');
const promBundle = require('express-prom-bundle');
const { Handler } = require('./handlers');
const { writeError } = require('./utils');
const { verifyMessage, id } = require('ethers/lib/utils');
const { constants } = require('ethers');

if (process.env.SENTRY_DSN) {
  console.log('initializing sentry');
  Sentry.init({
    dsn: process.env.SENTRY_DSN,
  });
}

const ALLOWED_METHODS = ['eth_sendBundle'];

function help() {
  console.log('node ./server/main.js minerUrls [PORT]');
}

function validPort(port) {
  if (isNaN(port) || port < 0 || port > 65535) {
    return false;
  }

  return true;
}

if (_.includes(process.argv, '-h') || _.includes(process.argv, '--help')) {
  help();
  process.exit(0);
}

let MINERS = [];
if (process.argv[2]) {
  MINERS = _.split(process.argv[2], ',');
}

const PORT = parseInt(_.get(process.argv, '[3]', '18545'));

if (!validPort(PORT)) {
  console.error(`invalid port specified for PORT: ${PORT}`);
  process.exit(1);
}

const app = express();
app.set('trust proxy', true);

app.use(
  rateLimit({
    windowMs: 60 * 1000, // 1 minute
    max: 60, // limit each IP to this many requests
    onLimitReached: (req) => {
      console.log(`rate limit reached for IP: ${req.ip}`);
    },
  }),
);
const metricsRequestMiddleware = promBundle({
  includePath: true,
  includeMethod: true,
  autoregister: false, // Do not register /metrics
  promClient: {
    collectDefaultMetrics: {},
  },
});
const { promClient, metricsMiddleware } = metricsRequestMiddleware;

// Metrics app to expose /metrics endpoint
const metricsApp = express();
metricsApp.use(metricsMiddleware);

const UNIQUE_USER_COUNT = {};
const bundleCounterPerUser = new promClient.Counter({
  name: 'relay_bundles_per_user',
  help: '# of bundles received per user',
  labelNames: ['username'],
});

app.use(metricsRequestMiddleware);
app.use(morgan('combined'));
function rawBodySaver(req, _, buf, encoding) {
  if (buf && buf.length) {
    req.rawBody = buf.toString(encoding || 'utf8');
  }
}
app.use(bodyParser.json({ verify: rawBodySaver }));
app.use(async (req, res, next) => {
  const auth = req.header('Authorization');
  let signature = req.header('X-Flashbots-Signature');
  if (!signature) {
    writeError(res, 403, 'missing X-Flashbots-Signature header');
    return;
  }
  if (!req.body) {
    writeError(res, 400, 'invalid json body');
    return;
  }
  if (!req.body.method) {
    writeError(res, 400, 'missing method');
    return;
  }
  if (!_.includes(ALLOWED_METHODS, req.body.method)) {
    writeError(
      res,
      400,
      `invalid method, only ${ALLOWED_METHODS} supported, you provided: ${req.body.method}`,
    );
    return;
  }

  req.user = {};
  if (signature) {
    signature = _.split(signature, ':');
    if (signature.length !== 2) {
      writeError(res, 403, 'invalid X-Flashbots-Signature header');
      return;
    }

    const msg = id(req.rawBody);
    try {
      const address = verifyMessage(msg, signature[1]);
      if (address === constants.AddressZero) {
        writeError(res, 403, `invalid signature for ${signature[0]}`);
        return;
      }
      if (address.toLowerCase() !== signature[0].toLowerCase()) {
        writeError(
          res,
          403,
          `signer address does not equal expected, got ${address}, expected ${signature[0]}`,
        );
        return;
      }
      req.user.address = address;
    } catch (err) {
      console.error(`error verifyMessage: ${err}`);
      writeError(res, 403, `error in signature check ${signature[0]}`);
      return;
    }
  }

  if (auth) {
    writeError(
      res,
      403,
      'invalid authorization method, API keys have been deprecated, use X-Flashbots-Signature',
    );
    return;
  }

  const username = req.user.address || req.user.keyID.slice(0, 8);
  let count = UNIQUE_USER_COUNT[username];
  if (!count) {
    count = 0;
  }
  UNIQUE_USER_COUNT[username] = count + 1;

  bundleCounterPerUser.inc({ username });
  next();
});

const handler = new Handler(MINERS, promClient);

app.use(async (req, res) => {
  try {
    if (req.body.method === 'eth_sendBundle') {
      await handler.handleSendBundle(req, res);
    } else if (req.body.method === 'eth_callBundle') {
      await handler.handleCallBundle(req, res);
    } else if (req.body.method === 'flashbots_getUserStats') {
      await handler.handleUserStats(req, res);
    } else {
      const err = `unknown method: ${req.body.method}`;
      Sentry.captureException(err);
      console.error(err);
      writeError(res, 400, err);
    }
  } catch (error) {
    Sentry.captureException(error);
    console.error(`error in handler: ${error}`);
    try {
      writeError(res, 500, 'internal server error');
    } catch (error2) {
      Sentry.captureException(error2);
      console.error(`error in error response: ${error2}`);
    }
  }
});
process.on('unhandledRejection', (err) => {
  Sentry.captureException(err);
  console.error(`unhandled rejection: ${err} ${err.stack}`);
});

app.listen(PORT, () => {
  metricsApp.listen(9090);

  console.log(`relay listening at ${PORT}`);
});
